feat: blog on using amazon s3 tables with Kestra (#2381)
kestra-io/docs: content/blogs/2025-03-27-using-amazon-s3-tables-with-kestra.md

---
title: "Using Amazon S3 Tables with Kestra"
description: "A step-by-step walkthrough of how we can orchestrate data loading into Amazon S3 tables using Kestra."
date: 2025-03-27T17:00:00
category: Solutions
author:
  name: Shruti Mantri
  image: smantri
  role: 
---

Amazon recently introduced S3 Tables, purpose-built for storing and querying tabular data directly on S3. Backed by built-in Apache Iceberg support, S3 Tables make data instantly accessible to popular AWS and third-party analytics engines like EMR and Athena.

> In this guide, we’ll show you how to orchestrate a complete workflow using Kestra—from downloading raw CSV files to converting them, uploading to S3, and creating Iceberg-backed S3 Tables. You’ll also learn how to query the data using Athena.

Reviewer: ```suggestion
In this post, we’ll show you how to orchestrate a complete workflow using Kestra—from downloading raw CSV files to converting them, uploading to S3, and creating Iceberg-backed S3 Tables. You’ll also learn how to query the data using Athena.
```

With Kestra, you can automate this entire process. Our declarative workflows handle data conversion, orchestration logic, infrastructure interactions, and job submission to EMR—all in a repeatable and trackable way.

This walkthrough will help you:

- Ingest and convert data into Parquet format

- Upload the data into S3

- Create an Iceberg table backed by S3 Table Buckets

- Automate querying through Athena

## Prerequisites

We will need the following as a prerequisite before we could proceed with the workflow creation:

- Kestra server running on version >= 0.21.0
- AWS account with access to IAM, S3 and Athena

Ensure you select the AWS region in which S3 Tables are [supported](https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-tables-regions-quotas.html).

## Implementing Kestra Workflow for S3 Tables

In order to get the end-to-end Kestra workflow that interacts with S3 table, we will need a lot of prework to be done. We will go through each of the step in detail. This will involve making some changes on the AWS console. In the process, we will also develop the Kestra workflow incrementally.

### Create general purpose S3 bucket

> Firstly, we will need a general purpose S3 bucket where we will store the data. For this, navigate to the [S3 service](https://console.aws.amazon.com/s3/home) on the AWS console. From the left navigation menu, select `General purpose buckets`. On the `General purpose buckets` page, click on the `Create bucket` button. On the `Create bukcet` page, provide a globally unique bucket name in the `Bucket name` text box. For the purpose of this blog, say we put the bucket name as `s3-general-purpose-ecommerce`. Rest of the configurations can be left as default, and click on the **Create bucket** button at the bottom of the page. This will create the new bucket.

Reviewer: ```suggestion
Firstly, we will need a general purpose S3 bucket where we will store the data. For this, navigate to the [S3 service](https://console.aws.amazon.com/s3/home) on the AWS console. From the left navigation menu, select `General purpose buckets`. On the `General purpose buckets` page, select the `Create bucket` button. On the `Create bucket` page, provide a globally unique bucket name in the `Bucket name` text box. For the purpose of this blog, for example we name the bucket as `s3-general-purpose-ecommerce`. Rest of the configurations can be left as default, and select the **Create bucket** button at the bottom of the page. This will create the new bucket.
```

![Create S3 General Purpose Bucket](/blogs/2025-03-27-using-amazon-s3-tables-with-kestra/create_s3_general_purpose_bucket.png)

> From the `General purpose bukcets` page, search for the newly created bucket, and click on the bucket name. On the corresponding bucket's home page, click on the `Create folder` button. Provide the folder name, say `data`, and click on the `Create folder` button at the bottom of the page. We will be storing our data in this folder.

Reviewer: ```suggestion
From the `General purpose buckets` page, search for the newly created bucket, and select the bucket name. On the corresponding bucket's home page, select the `Create folder` button. Provide the folder name, for example `data`, and select the `Create folder` button at the bottom of the page. We will be storing our data in this folder.
```

    uri: https://huggingface.co/datasets/kestra/datasets/raw/main/csv/products.csv

  - id: csv_to_ion
    type: io.kestra.plugin.serdes.csv.CsvToIon
    from: "{{ outputs.http_download.uri }}"

  - id: ion_to_parquet
    type: io.kestra.plugin.serdes.parquet.IonToParquet
    from: "{{ outputs.csv_to_ion.uri }}"
    schema: |
      {
        "type": "record",
        "name": "Product",
        "namespace": "com.kestra.product",
        "fields": [
          {"name": "product_id", "type": "int"},
          {"name": "product_name", "type": "string"},
          {"name": "product_category", "type": "string"},
          {"name": "brand", "type": "string"}
        ]
      }

  - id: s3_upload
    type: io.kestra.plugin.aws.s3.Upload
    accessKeyId: "{{ secret('AWS_ACCESS_KEY') }}"
    secretKeyId: "{{ secret('AWS_SECRET_KEY') }}"
    region: "{{ secret('AWS_REGION') }}"
    from: "{{ outputs.ion_to_parquet.uri }}"
    bucket: "s3-general-purpose-ecommerce"
    key: "data/products.parquet"
```

### Creating S3 Table Bucket

On the [S3 service](https://console.aws.amazon.com/s3/home) page on the AWS console, navigate to `Table buckets` from the left navigation menu.

> In case, you are opening the table buckets page for the first time, you will see a box at the top of the page about `Integration with AWS analytics services - New` with the `Enable integration` button. Click on the button and enable the integration of S3 table buckets with AWS analytics services like Amazon EMR, Amazon Redshift and Amazon Athena.

Reviewer: ```suggestion
In case, you are opening the table buckets page for the first time, you will see a box at the top of the page about `Integration with AWS analytics services - New` with the `Enable integration` button. Select the button and enable the integration of S3 table buckets with AWS analytics services like Amazon EMR, Amazon Redshift and Amazon Athena.
```

![Enable Integration](/blogs/2025-03-27-using-amazon-s3-tables-with-kestra/enable_integration.png)

> Next, we will create a table bucket. Click on the `Create table bucket` button at the top of the page. On the `Create table bucket` page, provide an appropriate name for the table bucket, say `ecommerce-lakehouse`. Click on the `Create table bucket` button at the bottom of the page. This will create the new table bucket.

Reviewer: ```suggestion
Next, we will create a table bucket. Select the `Create table bucket` button at the top of the page. On the `Create table bucket` page, provide an appropriate name for the table bucket, say `ecommerce-lakehouse`. Select the `Create table bucket` button at the bottom of the page. This will create the new table bucket.
```

![Create Table Bucket](/blogs/2025-03-27-using-amazon-s3-tables-with-kestra/create_table_bucket.png)

### Providing IAM Access

Navigate to [IAM service](https://console.aws.amazon.com/iam/home) on the AWS console. Navigate to `Roles` from the left navigation menu. Using the search box on the top of the page, ensure that you have `EMR_DefualtRole` and `EMR_EC2_DefaultRole` roles already present. If the roles are missing, you can create these default roles using `create-default-roles` as described [here](https://docs.aws.amazon.com/cli/latest/reference/emr/create-default-roles.html). Both these roles will be required for creating the EMR cluster.

> On the `Roles` page with IAM, search for `EMR_EC2_DefaultRole`, and click the same. On the `EMR_EC2_DefaultRole` role page, click on the `Add permissions` button, and from the dropdown that appears, click on `Attach Policies`. On the `Attach policy for EMR_EC2_DefaultRole` page, search for `AmazonS3TablesFullAccess`, select the `AmazonS3TablesFullAccess` policy and click on `Add permissions` button. This provides full access to S3 tables from the EC2 machines of the EMR.

Reviewer: ```suggestion
On the `Roles` page with IAM, search for `EMR_EC2_DefaultRole`, and select the same. On the `EMR_EC2_DefaultRole` role page, select the `Add permissions` button, and from the dropdown that appears, select `Attach Policies`. On the `Attach policy for EMR_EC2_DefaultRole` page, search for `AmazonS3TablesFullAccess`, select the `AmazonS3TablesFullAccess` policy and select `Add permissions` button. This provides full access to S3 tables from the EC2 machines of the EMR.
```

### Creating EC2 key pair

> Navigate to [EC2 service](https://console.aws.amazon.com/ec2/home) on the AWS console. From the left navigation menu, navigate to `Key pairs`. On the `Key pairs` page, click on the `Create key pair` button at the top of the page. On the `Create key pair` page, provide an appropriate name for the key pair, say `emr-ec2-key-pair`, and click on the `Create key pair` button at the bottom of the page. This will download the pem file associated with the key pair to your machine, and the new key pair will be created.

Reviewer: ```suggestion
Navigate to [EC2 service](https://console.aws.amazon.com/ec2/home) on the AWS console. From the left navigation menu, navigate to `Key pairs`. On the `Key pairs` page, select the `Create key pair` button at the top of the page. On the `Create key pair` page, provide an appropriate name for the key pair, say `emr-ec2-key-pair`, and select the `Create key pair` button at the bottom of the page. This will download the pem file associated with the key pair to your machine, and the new key pair will be created.
```

    region: "{{ secret('AWS_REGION') }}"
    clusterName: "Spark job cluster"
    logUri: "s3://s3-general-purpose-ecommerce/test-emr-logs"
    keepJobFlowAliveWhenNoSteps: true
    applications:
        - Spark
    masterInstanceType: m5.xlarge
    slaveInstanceType: m5.xlarge
    instanceCount: 3
    ec2KeyName: smantri-test
    releaseLabel: emr-7.5.0
    steps:
      - name: load_to_iceberg
        jar: "command-runner.jar"
        actionOnFailure: CONTINUE
        commands:
          - spark-submit --packages org.apache.iceberg:iceberg-spark-runtime-3.5_2.12:1.6.1,software.amazon.s3tables:s3-tables-catalog-for-iceberg:0.1.0 --conf spark.sql.catalog.s3tablesbucket=org.apache.iceberg.spark.SparkCatalog --conf spark.sql.catalog.s3tablesbucket.catalog-impl=software.amazon.s3tables.iceberg.S3TablesCatalog --conf spark.sql.catalog.s3tablesbucket.warehouse=arn:aws:s3tables:ap-southeast-1:1234567890:bucket/ecommerce-lakehouse --conf spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions s3://s3-general-purpose-ecommerce/load_to_iceberg.py --input s3://s3-general-purpose-ecommerce/data/products.parquet
    wait: true
```

> Running this workflow will get the data loaded onto S3 table bucket which can then be queried using Iceberg table in query services like Amazon Athena.

Reviewer: ```suggestion
Running this workflow will get the data loaded onto the S3 table bucket which can then be queried using Iceberg table in query services like Amazon Athena.
```

On the Query tab, you can write the query to get the data from this table:

```
SELECT * FROM "data"."products";
```

You should be able to see all the 20 rows from the table getting displayed in the `Query results` section.

![Query S3 Table using Amazon Athena](/blogs/2025-03-27-using-amazon-s3-tables-with-kestra/query_s3_table_using_amazon_athena.png)

Thus, we have successfully leverage S3 table bucket to create an Iceberg table.

## Conclusion

Kestra workflows can be used to work with the S3 table buckets and create Iceberge tables. This enables us to introduce orchestration related to S3 tables with Kestra, bringing in all the good features that S3 tables have to offer.

If you have any questions, reach out via [Kestra Community Slack](https://kestra.io/slack) or open [a GitHub issue](https://github.com/kestra-io/kestra). 

> If you like the project, give us [a GitHub star](https://github.com/kestra-io/kestra) and join [the open-source community](https://kestra.io/slack).

Reviewer: ```suggestion
If you like the project, give us a [GitHub star](https://github.com/kestra-io/kestra) and join [the open-source community](https://kestra.io/slack).
```
